### ISSUE TITLE
merge_TOAs() fails with a useless traceback

### ISSUE THREAD

**Issue by gooddc** [nanograv/PINT#1397]
This is coming from a mix of excised/not-excised tims, but the PINT problem twofold:
1. It would be great if merge_TOAs() could auto-fix this somehow, add a blank column or something and 
2. When merge_TOAs() fails, it doesn't provide anything useful for the user to fix the problem. 

The error message is just

> File /opt/conda/lib/python3.9/site-packages/pint/toa.py:263, in get_TOAs(timfile, ephem, include_bipm, bipm_version, include_gps, planets, model, usepickle, tdb_method, picklefilename, limits)
>     261     t = TOAs(timfile)
>     262 else:
> --> 263     t = merge_TOAs([TOAs(t) for t in timfile])
>     265 files = [t.filename] if isinstance(t.filename, (str, Path)) else t.filename
>     266 if files is not None:
> 
> File /opt/conda/lib/python3.9/site-packages/pint/toa.py:2420, in merge_TOAs(TOAs_list)
>    2418 num_cols = [len(tt.table.columns) for tt in TOAs_list]
>    2419 if len(set(num_cols)) > 1:
> -> 2420     raise TypeError(
>    2421         f"merge_TOAs() cannot merge. Inconsistent numbers of table columns: {num_cols}"
>    2422     )
>    2423 # Use a copy of the first TOAs instance as the base for the joined object
>    2424 nt = copy.deepcopy(TOAs_list[0])
> 
> TypeError: merge_TOAs() cannot merge. Inconsistent numbers of table columns: [9, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8]

I'm not even sure what the 8 columns are in the tims that aren't causing the problem, so it's hard to track down what's different between the two.

**Comment by dlakaplan** [comment]
Could you provide example files/command that cause this error?

**Comment by paulray** [comment]
My table has more than 9 columns. Make sure to compute TDBs and posvels.
```
In [3]: ts = pint.toa.get_TOAs("NGC6440E.tim")
Downloading https://raw.githubusercontent.com/ipta/pulsar-clock-corrections/main/index.txt
|=====================================================================================================| 3.4k/3.4k (100.00%)         0s

In [4]: ts.table.colnames
Out[4]: 
['index',
 'mjd',
 'mjd_float',
 'error',
 'freq',
 'obs',
 'flags',
 'delta_pulse_number',
 'tdb',
 'tdbld',
 'ssb_obs_pos',
 'ssb_obs_vel',
 'obs_sun_pos']
```

**Comment by gooddc** [comment]
Here's my "minimal unworking" example. 
[unworking_example.zip](https://github.com/nanograv/PINT/files/9568831/unworking_example.zip)

**Comment by dlakaplan** [comment]
Is there a version of that you could do without `timing_analysis`?

**Comment by dlakaplan** [comment]
So the issue is that the NANOGrav files have pulse numbers and the others don't.  It's straightforward to delete that before writing (`del t.table['pulse_number']` or similar).  We could also add options to force TOA writing to not write that out, or the same with reading.  

Making the error message more useful is harder, since when you are reading in a bunch of TOAs at once it's a little complicated to only give the differences.  I'll try something to start.

**Comment by dlakaplan** [comment]
#1398 may solve this.  You can change your TOA reading to use `include_pn=False` and it won't read in the pulse numbers and so won't have the conflicts.  It may still be an issue (do you really want to throw those out?) but not a coding one.

@paulray : the reason for the small number of columns (I think) is that if you try to read in all of the files at once (like `get_TOAs([file1,file2,...])`) it tries to merge before it does the rest of the calculations.

**Comment by scottransom** [comment]
TBH, I'm not sure why we are saving pulse numbers in our text-based .tim files.  That seems completely unnecessary since we are providing our parfiles, as well, and the pulse numbers can be re-computed with a single quite quick command.  So maybe for our final "released" .tim files we remove the pn flags?  Thoughts @swiggumj?

**Comment by aarchiba** [comment]
> TBH, I'm not sure why we are saving pulse numbers in our text-based .tim files. That seems completely unnecessary since we are providing our parfiles, as well, and the pulse numbers can be re-computed with a single quite quick command. So maybe for our final "released" .tim files we remove the pn flags? Thoughts @swiggumj?

I think pulse numbers are extremely useful information to pass in to further fitting tools, and they are easily ignored if not needed.

**Comment by aarchiba** [comment]
> So the issue is that the NANOGrav files have pulse numbers and the others don't. It's straightforward to delete that before writing (`del t.table['pulse_number']` or similar). We could also add options to force TOA writing to not write that out, or the same with reading.
> 
> Making the error message more useful is harder, since when you are reading in a bunch of TOAs at once it's a little complicated to only give the differences. I'll try something to start.

I think `merge_TOAs` should record "unknown pulse number" (NaN) for files that don't have pulse numbers, and then work.

**Comment by aarchiba** [comment]
> My table has more than 9 columns. Make sure to compute TDBs and posvels.
> 
> ```
> In [3]: ts = pint.toa.get_TOAs("NGC6440E.tim")
> Downloading https://raw.githubusercontent.com/ipta/pulsar-clock-corrections/main/index.txt
> |=====================================================================================================| 3.4k/3.4k (100.00%)         0s
> 
> In [4]: ts.table.colnames
> Out[4]: 
> ['index',
>  'mjd',
>  'mjd_float',
>  'error',
>  'freq',
>  'obs',
>  'flags',
>  'delta_pulse_number',
>  'tdb',
>  'tdbld',
>  'ssb_obs_pos',
>  'ssb_obs_vel',
>  'obs_sun_pos']
> ```

`merge_TOAs` can and should report which column names are present in some but not all TOAs objects; this would make diagnosing this situation very much easier.

**Comment by dlakaplan** [comment]
Fixed by #1398
